craft_grammar/models.py

```diff
@@ -50,10 +53,77 @@ def _grammar_append(cls, entry: List, item: Any) -> None:
             _mark_and_append(entry, {key: cls.validate(value)})
 
 
-_GrammarType = Dict[str, Any]
+# Public types for grammar-enabled attributes
+class GrammarBool(_GrammarBase):
```

Comment:
This repetition kind of bothers me because these classes are almost identical, _plus_ they are actually more restrictive than the models (e.g. Parts accept the float `1.0` for `source-tag`, because after validation Pydantic stringifies it).

However I don't think it's worth changing this now, and we can always do it later if we decide not to move to a generic `Grammar[T]` type.

Reply:
Hmmm, `source-tag` should be a str IMO. The Pydantic maybe should do the convert.

Reply:
oh I just meant that `source-tag`'s type is `str`, but a yaml like `source-tag: 1.0` will pass Part validation but not GrammarStr validation